proposals/stdlib/base64.md

```diff
@@ -0,0 +1,313 @@
+* `Base64.Default` and `Base64.UrlSafe` do not separate the result into multiple lines.
+* `Base64.Mime` adds a CRLF every 76 symbols, but it does not add a CRLF at the end of the encoded output.
+
+For decode operations:
+* Whether a `Base64` instance requires, prohibits, or allows padding in the input symbols is determined by the
+  padding option set for the instance. The padding option for all predefined instances is set to `PaddingOption.PRESENT`.
+* All `Base64` instances interpret the padding character as the end of the encoded binary data, and subsequent symbols are prohibited.
+* All `Base64` instances require the pad bits to be zeros.
+* `Base64.Default` throws if it encounters a character outside "The Base64 Alphabet" as specified in Table 1 of RFC 4648.
+* `Base64.UrlSafe` throws if it encounters a character outside "The URL and Filename safe Base 64 Alphabet" as specified in Table 2 of RFC 4648.
+* `Base64.Mime` ignores all line separators and other characters outside "The Base64 Alphabet" as specified in Table 1 of RFC 2045.
+
+### Examples
+
+This is an example of how the proposed Base64 API could be used:
+
+```kotlin
+val foBytes = "fo".map { it.code.toByte() }.toByteArray()
+Base64.Default.encode(foBytes) // "Zm8="
+// Alternatively:
+// Base64.encode(foBytes)
+
+Base64.withPadding(Base64.PaddingOption.ABSENT).encode(foBytes) // "Zm8"
+
+val foobarBytes = "foobar".map { it.code.toByte() }.toByteArray()
+Base64.UrlSafe.encode(foobarBytes) // "Zm9vYmFy"
+
+Base64.Default.decode("Zm8=") // foBytes
+// Alternatively:
+// Base64.decode("Zm8=")
+
+Base64.UrlSafe.decode("Zm9vYmFy") // foobarBytes
+```
+
+## Alternatives
+
+Because Kotlin didn't provide a Common API for Base64 encoding, community has implemented a few libraries targeting this gap:
+* https://github.com/05nelsonm/encoding
+* https://github.com/saschpe/Kase64
+
+Providing Base64 encoding/decoding directly in the Kotlin standard library avoids additional dependencies,
+ensures the API is consistent with other Kotlin stdlib API, and ensures it has a good performance.
+
+## Naming
+
+### Existing functions for encoding a String (UTF-16) to UTF-8 and decoding back
+
+```kotlin
+public fun String.encodeToByteArray(...): ByteArray
+
+public fun ByteArray.decodeToString(...): String
+```
+
+### Proposed naming
+
+When researching usages of Java `Base64`, we observed two patterns that influenced our design and naming.
+First, the vast majority of users obtain an encoder or decoder and use it immediately,
+e.g., `Base64.getDecoder().decode(...)`. This led us to avoid separating the encoder and decoder into different entities.
+We also didn't find evidence that the benefits of such a separation outweigh the usage inconvenience. 
+Second, encoding to `String` and decoding from `String` are considerably more popular than encoding to `ByteArray` and
+decoding from `ByteArray`. Thus, we decided to make the more frequently used functions shorter.
```

Comment:
Another reason was that a compound operation like "encode bytearray to base64 string + encode string to utf8 byte array" should have a compound name.

Reply:
The reverse operation (utf8 byte array -> string -> byte array) doesn't have a compound name.

Reply:
Yes, in that case we didn't complicate the name thanks to overloading on parameter types.

Reply:
Done!